File: ProblemSolve/560.cpp
/**************************
 * Subarray Sums Eqauls K *
 **************************/

//참고 : https://leetcode.com/problems/subarray-sum-equals-k/solution/

class Solution {
public:
    int subarraySum(vector<int>& nums, int k) {
        int len = nums.size();
        int sum = 0;
        int ans = 0;
        unordered_map<int, int> map;
        map[0] = 1;
        
        for(int i=0; i<len; i++){
            sum += nums[i];
                        
            if(map.find(sum - k) != map.end()){
                ans += map[sum - k];
            }
            
            map[sum]++;
        }
        
        return ans;
    }
};

File: ProblemSolve/560.cs
/********************************
 * Subarray Sums Divisible by K *
 ********************************/

public class Solution {
    public int SubarraySum(int[] nums, int k) {
        int sum = 0;
        int ans = 0;
        Dictionary<int, int> hash = new Dictionary<int, int>(){{0, 1}};
        
        foreach(int num in nums){
            sum += num;
            
            if(hash.ContainsKey(sum - k)){
                ans += hash[sum - k];
            }
            
            if(hash.ContainsKey(sum)){
                hash[sum]++;
            }
            else{
                hash[sum] = 1;
            }            
        }
        
        return ans;
    }
}